feat(components): initial mutations over time component

JonasKellerer opened the pull request:
<!-- Mention the issue that this PR resolves. Delete if there is no corresponding issue -->
resolves #59

### Summary
<!-- 
Add a few sentences describing the main changes introduced in this PR
This is only relevant, if there is no issue that already contains the information
-->
Adds a new component, which shows the mutations over time, similar to covSpectrum. 
Currently missing are the tooltips and the selector for the color scale. They will be done in #353 and #354.

### Screenshot
<!-- 
When applicable, add a screenshot showing the main effect this PR has, even if "trivial":
e.g. changed layout, changed button text, different logging in backend.
This helps others quickly grasping what you did even if they are not familiar with the code base.
-->

![grafik](https://github.com/user-attachments/assets/46a00d2f-5bbb-4ec9-ba4c-6c88ff2cb3e4)


### PR Checklist
<!-- Check completed items of strikethrough irrelevant items (using ~~text~~) -->
- [x] All necessary documentation has been adapted.
- [x] The implemented feature is covered by an appropriate test.

JonasKellerer commented:
This PR is missing a test for the queryMutationOverTime function, which will be added.

chaoran-chen reviewed (commented):
Amazing!

I briefly tested it and found a few small issues:

- The granularity switch doesn't work for me.
- Let's better call it "mutation**s** over time", not "mutation" (plural!)
- The code example in storybooks uses "<gs-mutations" without "-over-time"

JonasKellerer commented on components/src/preact/mutationsOverTime/mutations-over-time.tsx:
```diff
@@ -0,0 +1,175 @@
+                <CsvDownloadButton
+                    className='mx-1 btn btn-xs'
+                    getData={() => {
+                        return [{ value: 1 }, { value: 2 }];
```
There is now an issue #362

JonasKellerer commented on components/src/preact/mutationsOverTime/mutations-over-time.tsx:
```diff
@@ -0,0 +1,175 @@
+                    getData={() => {
+                        return [{ value: 1 }, { value: 2 }];
+                    }}
+                    filename='mutation-over-time.csv'
```
This will be done in #362

JonasKellerer commented on components/src/web-components/visualization/gs-mutations-over-time.stories.ts:
```diff
@@ -0,0 +1,214 @@
+    lapisFilter='{ "pangoLineage": "JN.1*", "dateFrom": "2024-01-15", "dateTo": "2024-07-10" }'
+    sequenceType="nucleotide"
+    views='["grid"]'
+    headline="Mutations over time"
```
Was removed during rebase.

JonasKellerer commented on components/src/web-components/visualization/gs-mutations-over-time.tsx:
```diff
@@ -0,0 +1,90 @@
+/* eslint-disable @typescript-eslint/no-unused-vars, no-unused-vars */
+type LapisFilterMatches = Expect<Equals<typeof MutationsOverTimeComponent.prototype.lapisFilter, LapisFilter>>;
+type SequenceTypeMatches = Expect<Equals<typeof MutationsOverTimeComponent.prototype.sequenceType, SequenceType>>;
+type ViewsMatches = Expect<Equals<typeof MutationsOverTimeComponent.prototype.views, View[]>>;
```
I changed it here, but added another ticket for all the other components, which have the same style:  #363

fengelniederhammer reviewed (changes requested):
Some minor things. In general it looks good :+1:

fengelniederhammer commented:
It would also be good to link the issue in the commit message (either appending the issue to the header or by using `closes #` or `Refs: #` in the footer.

fengelniederhammer commented on components/vitest.setup.ts:
```diff
@@ -69,6 +69,41 @@ export const lapisRequestMocks = {
+        sequenceType: 'nucleotide' | 'amino acid',
+    ) => {
+        testServer.use(
+            http.post(substitutionsOrDeletionsEndpoint(DUMMY_LAPIS_URL, sequenceType), async ({ request }) => {
```
Probably `resolver` could even use `any` instead of the generics.

JonasKellerer commented on components/src/query/queryMutationsOverTime.spec.ts:
```diff
@@ -0,0 +1,282 @@
+
+        const result = await queryMutationsOverTimeData(lapisFilter, 'nucleotide', DUMMY_LAPIS_URL, dateField, 'day');
+
+        expect(result.getAsArray(0)).to.deep.equal([
+            [0.1, 0.3, 0],
+            [0.4, 0, 0],
+        ]);
```
Yeah. I was also wandering about it. My reasoning for not sorting it is that a regular map also does not sort it. So the user is responsible for sorting. In this case the sorting of the date keys in the grid component.

fengelniederhammer reviewed (approved):
LGTM

* Do we still need to implement a limit on how many mutations we show in the grid at a time?
* Do we need the intermediate commit that sets the minProportion? AFAIK it only affects this (new) component anyway.

JonasKellerer commented:
As discussed: we added a new ticket #367 to limit the number of shown mutations/dates. For now we keep the component as is.

I rebased the PR, so we get rid of the extra commit.
